### ISSUE TITLE
Offline capability for Eddystone URLs

### ISSUE THREAD

**Issue by schildbach** [google/physical-web#738]
Currently it seems as if URLs are always requested before being displayed as a notification. Only if the HTTP request succeeds a notification is shown. That severely limits usecases; in particular you need to have an Internet connection. Which actually in many cases you haven't (subway, stores in basements, big concrete buildings, etc.).

I think it would be useful if an URL from a beacon would be resolved via the Intent resolver first. If there is an app that can take your intent (e.g. via intent-filter in the AndroidManifest) just show the notification to the user and let the user pass the URL into that app via the standard Intent chooser mechanism. Chances are the app can properly use the URL without internet access, e.g. because the data is already in the app. By the way this is how NFC works more or less with NDEF messages of type URL.

**Comment by cco3** [comment]
Right, we are investigating several solutions to address this issue.  Some difficult problems include:
* Displaying a title and description
* Communicating to the user that the URL hasn't been passed through a filter (i.e. something like safesearch)

**Comment by schildbach** [comment]
Imho if the user has installed an app that can consume the URL this is a strong indication that the user wants to see the notification.

The title can be acquired from (one of) the apps it resolves to simply by using the name of the matched intent filter (that's what the intent chooser does, too).

**Comment by cco3** [comment]
Ah, I see what you are saying.  Yes, we've discussed this but not implemented it.  It's a good idea.

**Comment by schildbach** [comment]
@scottjenson Why did you close this without comment?

**Comment by scottjenson** [comment]
Sorry! Given that
1) we've opened another issue on FatBeacons, which addresses this issue and
2) there wasn't much additional activity

I assumed it was safe to close. If there is something else you'd like to add, please do.

**Comment by schildbach** [comment]
Hmm, looks like FatBeacon is an entirely different idea. This ticket is about how to associate apps to URLs in an offline way.

**Comment by scottjenson** [comment]
OK, it appears I was triaging a bit too argressively this morning. let's take this back up and get it resolved. There appear to be two issues here: displaying then using/opening the URL. To display it, we need to get the metadata for the URL: title, description and favicon. That currently requires a trip to the web to find this page. This information is fetched by our PWS (Physical Web Service) to keep the user from being tracked. 

What I understand from your proposal is how the URL is used/opened. Is that correct? How is what you proposing different from just using the standard Android Intent model (whichever intent wants the URL gets it) This allows maps URLs be openned by maps and unclaimed URLs opened by the default web browser.

**Comment by schildbach** [comment]
My idea is to actually *use* the standard intent model, rather than implementing something proprietary that requires internet access. Like I said: Have a look like its done in the NFC API, it can do entirely without internet access.

**Comment by scottjenson** [comment]
I don't disagree with you, I just don't understand you. Let's take my first question: how would you show Title/Description/Favicon without internet access?

**Comment by schildbach** [comment]
When using the Android intent chooser, the matching apps provide a description and an icon via their intent filters.

**Comment by scottjenson** [comment]
The issue I see with that is that we'll mostly have URLs that point to web pages. How should we show both Web and Intent links to users at the same time? I don't see how we can mix the intent choose items into the web items.

**Comment by schildbach** [comment]
The Android intent chooser is only using public API. You can use the same API to get all the suggestion, then mix it with the "web links" and use your own UI.

**Comment by mmocny** [comment]
This is a neat idea.  As Conley said up top, we have broadly discussed before but haven't implemented.  Thanks for offering a few concrete suggestions!

One caveat, I think we could only get generic app-specific metadata from the installed app, and not necessarily metadata about the specific deep link result.  It's probably still useful for a large swath of use cases and we should absolutely pursue it.

Another broad question: is there any way to know (or guess, within reason) that an app can open a link without hitting network?  Just because the app claims it can handle the intent, doesn't mean it can provide a useful result.  It could just be a custom web browser using a WebView, or require network calls to get result-specific data, or whatnot.

**Comment by schildbach** [comment]
@mmocny Apps can programmatically enable or disable intent filters. I think they could use this to show their filter only if internet is present. However I really think it's up to the apps to present something useful if no internet is present, e.g. from cached results or a local database.

**Comment by mmocny** [comment]
Good point, but we'll have to see just how many apps actually are so well behaved.

One other observation, which we should perhaps have mentioned earlier:  If the app is already installed, it is free to do its own beacon scanning.  Including it in PW/Nearby results is possibly useful, but there are existing viable solutions which app devs don't need to wait for.
